File: java/0198-house-robber.java
import java.util.HashMap;
import java.util.Map;

class Solution {
    public int rob(int[] nums) {
        int n = nums.length;
        Map<Integer, Integer> memo = new HashMap<>();
        return helper(n, 0, nums, memo);
    }

    private int helper(int n, int i, int[] nums, Map<Integer, Integer> memo) {
        if (i >= n) return 0;
        if (memo.containsKey(i)) return memo.get(i);
        int res = Math.max(helper(n, i + 1, nums, memo), helper(n, i + 2, nums, memo) + nums[i]);
        memo.put(i, res);
        return res;
    }
}

File: 0198-house-robber.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        prev_max, curr_max = 0, 0
        for n in nums:
            prev_max, curr_max = curr_max, max(prev_max + n, curr_max)
        return curr_max


# class Solution:
#     def rob(self, nums: List[int]) -> int:
#         n = len(nums)
#         if n < 3:
#             return max(nums)
#         @cache
#         def helper(i):
#             if i >= n:
#                 return 0
#             return max(nums[i] + helper(i + 2), helper(i + 1))
#         return helper(0)

# class Solution:
#     def rob(self, nums: List[int]) -> int:
#         n = len(nums)
#         if n < 3:
#             return max(nums)
#         memo = {}
#         def helper(i):
#             if i >= n:
#                 return 0
#             if i in memo:
#                 return memo[i]
#             memo[i] =  max(nums[i] + helper(i + 2), helper(i + 1))
#             return memo[i]
#         return helper(0)
